File: MockInterview.Frontend/src/types/entityQueryOptions.ts
import {FilterOptions} from "./filterOptions";
import {IncludeOptions} from "./includeOptions";
import {SortOptions} from "./sortOptions";
import {PaginationOptions} from "./paginationOptions";
import {IMappable, Util} from "./common";

export interface IEntityQueryOptions {
    filterOptions: FilterOptions;
    includeOptions: IncludeOptions;
    sortOptions: SortOptions;
    paginationOptions: PaginationOptions;
}

export class EntityQueryOptions implements IMappable<IEntityQueryOptions> {
    filterOptions!: FilterOptions | null;
    includeOptions!: IncludeOptions | null;
    sortOptions!: SortOptions | null;
    paginationOptions!: PaginationOptions | null;

    mapFrom(data: IEntityQueryOptions) {
        this.filterOptions = Util.mapObject(data.filterOptions, FilterOptions);
        this.includeOptions = Util.mapObject(data.includeOptions, IncludeOptions);
        this.sortOptions = Util.mapObject(data.sortOptions, SortOptions);
        this.paginationOptions = Util.mapObject(data.paginationOptions, PaginationOptions);
    }
}

File: MockInterview.Backend/MockInterview.DAL/Models/Query/EntityQueryOptions.cs
using MockInterview.Core.Models.Common;

namespace MockInterview.DAL.Models.Query;

/// <summary>
/// Represents queryable entities source query options
/// </summary>
/// <typeparam name="TEntity"></typeparam>
public class EntityQueryOptions<TEntity> : IEntityQueryOptions<TEntity> where TEntity : class, IEntity
{
    public SearchOptions<TEntity>? SearchOptions { get; set; }
    
    public FilterOptions<TEntity>? FilterOptions { get; set; }
    
    public IncludeOptions<TEntity>? IncludeOptions { get; set; }
    
    public SortOptions? SortOptions { get; set; }
    
    public PaginationOptions PaginationOptions { get; set; } = null!;
}